src/ConstructiveSolidGeometry/VolumePrimitives/Cone.jl

```diff
@@ -1,207 +1,850 @@
+end
+function _in(pt::CartesianPoint, c::PartialVaryingTube{T,OpenPrimitive}) where {T} 
+    abs(pt.z) < c.hZ &&
+    radius_at_z(c.hZ, c.r[1][1], c.r[2][1], pt.z) < hypot(pt.x, pt.y) < radius_at_z(c.hZ, c.r[1][2], c.r[2][2], pt.z) &&
+    _in_angular_interval_open(atan(pt.y, pt.x), c.φ)
+end
 
-in(p::AbstractCoordinatePoint, c::Cone{<:Any, <:Any, Nothing, <:Any}) =
-    _in_z(p, c.z) && _in_cyl_r(p, get_r_at_z(c, p.z))
+function surfaces(t::PartialVaryingTube{T}) where {T}
+    bot_center_pt = _transform_into_global_coordinate_system(CartesianPoint{T}(zero(T), zero(T), -t.hZ), t) 
+    top_center_pt = _transform_into_global_coordinate_system(CartesianPoint{T}(zero(T), zero(T), +t.hZ), t) 
+    inner_mantle = ConeMantle{T,Tuple{T,T},Tuple{T,T},:outwards}((t.r[1][1], t.r[2][1]), t.φ, t.hZ, t.origin, t.rotation)
+    outer_mantle = ConeMantle{T,Tuple{T,T},Tuple{T,T},:inwards}( (t.r[1][2], t.r[2][2]), t.φ, t.hZ, t.origin, t.rotation)
+    e_bot = PartialAnnulus{T}(r = (t.r[1][1], t.r[1][2]), φ = t.φ, origin = bot_center_pt, rotation = t.rotation)
+    e_top = PartialAnnulus{T}(r = (t.r[2][1], t.r[2][2]), φ = t.φ, origin = top_center_pt, rotation = -t.rotation * RotZ{T}(π))
+    poly_l = _transform_into_global_coordinate_system(Quadrangle{T}((
+        CartesianPoint(CylindricalPoint{T}(t.r[1][1],  t.φ[1], -t.hZ)),
+        CartesianPoint(CylindricalPoint{T}(t.r[2][1],  t.φ[1], +t.hZ)),
+        CartesianPoint(CylindricalPoint{T}(t.r[2][2],  t.φ[1], +t.hZ)),
+        CartesianPoint(CylindricalPoint{T}(t.r[1][2],  t.φ[1], -t.hZ)) )), t)
+    poly_r = _transform_into_global_coordinate_system(Quadrangle{T}((
+        CartesianPoint(CylindricalPoint{T}(t.r[1][1],  t.φ[2], -t.hZ)),
+        CartesianPoint(CylindricalPoint{T}(t.r[1][2],  t.φ[2], -t.hZ)),
+        CartesianPoint(CylindricalPoint{T}(t.r[2][2],  t.φ[2], +t.hZ)),
+        CartesianPoint(CylindricalPoint{T}(t.r[2][1],  t.φ[2], +t.hZ)) )), t)
+    # normals of the surfaces show inside the volume primitives. 
+    e_top, e_bot, inner_mantle, outer_mantle, poly_l, poly_r
+end
 
-in(p::AbstractCoordinatePoint, c::Cone{<:Any, <:Any, <:AbstractInterval, <:Any}) =
-    _in_z(p, c.z) && _in_φ(p, c.φ) && _in_cyl_r(p, get_r_at_z(c, p.z))
+####################################################################
+####################################################################
 
-# read-in
-function Geometry(::Type{T}, t::Union{Type{Cone}, Type{Tube}}, dict::AbstractDict, input_units::NamedTuple) where {T}
-    length_unit = input_units.length
-    angle_unit = input_units.angle
-    r = parse_r_of_primitive(T, dict, length_unit)
-    φ = parse_φ_of_primitive(T, dict, angle_unit)
-    z = parse_height_of_primitive(T, dict, length_unit)
-    return Cone(T, r, φ, z)
+#=
+
+\
+ \
+__\
+
+=#
+const UpwardCone{T,CO} = Cone{T,CO,Tuple{Tuple{Nothing,T},Nothing},Nothing} # Full in φ
+const PartialUpwardCone{T,CO} = Cone{T,CO,Tuple{Tuple{Nothing,T},Nothing},Tuple{T,T}} 
```

Comment:
I guess that many of the methods for the primitives can be somewhat combined to reduce lines of code.

Reply:
Maybe. Feel free to reduce it.